src/branched_evolve.jl

```diff
@@ -0,0 +1,2050 @@
+using Quasar: QasmExpression, head, ClassicalVariable, GateDefinition, FunctionDefinition, SizedBitVector
+
+# Define a type to represent measurement values for each simulation path
+# path_outcomes takes path indices as keys with the measurement results
+# as dictionaries where the key is the qubit name and value is the qubit
+# measurement
+struct MeasurementValues
+    path_outcomes::Dict{Int, Dict{String, Int}}
+end
+
+###############################################
+# Helper functions for evaluating expressions #
+###############################################
+
+"""
+    evaluate_qubits(sim::BranchedSimulator, qubit_expr::QasmExpression) -> Dict{Int, Vector{Int}}
+
+Evaluate qubit expressions to get qubit indices. This replaces the visitor's evaluate_qubits function.
+Returns a dictionary mapping path indices to lists of qubit indices in the state.
+The same indices are used for all active simulations since we can't define qubits in a local scope.
+
+As of now, this only works with registers and classical bits. Functionality for any classical variable
+still needs to be implemented.
+"""
+function evaluate_qubits(sim::BranchedSimulator, qubit_expr::QasmExpression)
```

Comment:
This is a really long function; it would be easier to reason about if it were split into multiple helpers